README.md
# Restivus [v1.1.0](https://github.com/icellan/meteor-restivus/blob/devel/CHANGELOG.md#change-log)

#### REST APIs for the Best of Us!

**NOTE: This package is based on [nimble:restivus](https://github.com/kahmali/meteor-restivus) created by [Kahmali Rose](https://github.com/kahmali) but has been decaffeinated and expanded.**

Restivus makes building REST APIs in Meteor 1.10.0+ easier than ever before! The package is inspired
by [RestStop2][reststop2-docs] and [Collection API](https://github.com/crazytoad/meteor-collectionapi),
and is built on top of [Simple JSON Routes][json-routes] to provide:

- A simple interface for creating REST APIs
- Easy setup of CRUD endpoints for Mongo Collections
- User authentication via the API
  - Optional login and logout endpoints
  - Access to `this.user` in authenticated endpoints
  - Custom authentication if needed
- Role permissions for limiting access to specific endpoints
  - Works alongside the [`alanning:roles`][alanning-roles] package - Meteor's accepted role
    permission package
- API versioning via URL path
- And also planned:
  - JSON PATCH support on collections
  - Autogenerated OPTIONS endpoint on routes
  - Pre and post hooks on all endpoints

### [Version 1.1.0 Released!](#upgrading-to-110)

## Table of Contents

- [Getting Started](#getting-started)
  - [Installation](#installation)
  - [Quick Start](#quick-start)
  - [Terminology](#terminology)
- [Writing a Restivus API](#writing-a-restivus-api)
  - [Configuration Options](#configuration-options)
  - [Defining Collection Routes](#defining-collection-routes)
    - [Collection](#collection)
    - [Collection Options](#collection-options)
      - [Route Configuration](#route-configuration)
      - [Endpoint Configuration](#endpoint-configuration)
    - [Request and Response Structure](#request-and-response-structure)
    - [Users Collection Endpoints](#users-collection-endpoints)
  - [Defining Custom Routes](#defining-custom-routes)
    - [Path Structure](#path-structure)
    - [Route Options](#route-options)
    - [Defining Endpoints](#defining-endpoints)
      - [Endpoint Configuration](#endpoint-configuration-1)
      - [Endpoint Context](#endpoint-context)
      - [Response Data](#response-data)
  - [Versioning an API](#versioning-an-api)
  - [Documenting an API](#documenting-an-api)
- [Consuming a Restivus API](#consuming-a-restivus-api)
  - [Basic Usage](#basic-usage)
  - [Authenticating](#authenticating)
    - [Default Authentication](#default-authentication)
      - [Logging In](#logging-in)
      - [Logging Out](#logging-out)
      - [Authenticated Calls](#authenticated-calls)
- [Upgrading Restivus](#upgrading-restivus)
  - [Upgrading to 1.1.0](#upgrading-to-110)
  - [Upgrading to 1.0.0](#upgrading-to-100)
  - [Upgrading to 0.8.0](#upgrading-to-080)
  - [Upgrading to 0.7.0](#upgrading-to-070)
  - [Upgrading to 0.6.1](#upgrading-to-061)
- [Resources](#resources)
  - [Plugins](#plugins) 
  - [Change Log](#change-log)
  - [Contributing](#contributing)

# Getting Started

## Installation

You can install Restivus using Meteor's package manager:
```bash
> meteor add icellan:restivus
```

## Quick Start

Often, the easiest way to explain something is by example, so here's a short example of what it's
like to create an API with Restivus (keep scrolling for a JavaScript version):

###### JavaScript:
```javascript
Items = new Mongo.Collection('items');
Articles = new Mongo.Collection('articles');

if (Meteor.isServer) {
  // Global API configuration
  const Api = new Restivus({
    useDefaultAuth: true,
    prettyJson: true
  });

  // Generates: GET, POST on /api/items and GET, PUT, PATCH, DELETE on
  // /api/items/:id for the Items collection
  Api.addCollection(Items);

  // Generates: POST on /api/users and GET, DELETE /api/users/:id for
  // Meteor.users collection
  Api.addCollection(Meteor.users, {
    excludedEndpoints: ['getAll', 'put'],
    routeOptions: {
      authRequired: true
    },
    endpoints: {
      post: {
        authRequired: false
      },
      delete: {
        roleRequired: 'admin'
      }
    }
  });

  // Maps to: /api/articles/:id
  Api.addRoute('articles/:id', {authRequired: true}, {
    get: function () {
      return Articles.findOne(this.urlParams.id);
    },
    delete: {
      roleRequired: ['author', 'admin'],
      action: function () {
        if (Articles.remove(this.urlParams.id)) {
          return {status: 'success', data: {message: 'Article removed'}};
        }
        return {
          statusCode: 404,
          body: {status: 'fail', message: 'Article not found'}
        };
      }
    }
  });
}
```

## Terminology

Just to clarify some terminology that will be used throughout these docs:

_**(HTTP) Method:**_
- The type of HTTP request (e.g., GET, PUT, POST, etc.)

_**Endpoint:**_
- The function executed when a request is made at a given URL path for a specific HTTP method

_**Route:**_
- A URL path and its set of configurable endpoints

# Writing A Restivus API

**Restivus is a _server-only_ package. Attempting to access any of its methods from the client will
result in an error.**

## Configuration Options

The following configuration options are available when initializing an API using
`new Restivus(options)`:

##### `apiPath`
- _String_
- Default: `'api/'`
- The base path for your API. If you use `'api'` and add a route called `'users'`, the URL will be
  `https://yoursite.com/api/users/`.

##### `auth`
- _Object_
  - `token`  _String_
    - Default: `'services.resume.loginTokens.hashedToken'`
    - The path to the hashed auth token in the `Meteor.user` document. This location will be checked
      for a matching token if one is returned in `auth.user()`.
  - `user`  _Function_
    - Default: Get user ID and auth token from `X-User-Id` and `X-Auth-Token` headers
        ```javascript
        function() {
          return {
            userId: this.request.headers['x-user-id'],
            token: Accounts._hashLoginToken(this.request.headers['x-auth-token'])
          };
        }
        ```

    - Provides one of two levels of authentication, depending on the data returned. The context
      within this function is the [endpoint context](#endpoint-context) without `this.user` and
      `this.userId` (well, that's what we're working on here!). Once the user authentication
      completes successfully, the authenticated user and their ID will be attached to the [endpoint
      context](#endpoint-context). The two levels of custom authentication and their required return
      data are:
      - Partial auth
        - `userId`: The ID of the user being authenticated
        - `token`: The auth token to be verified
        - If both a `userId` and `token` are returned, Restivus will hash the token, then,
          authentication will succeed if the `hashedToken`
          exists in the given `Meteor.user` document at the location specified in `auth.token`
      - Complete auth
        - `user`: The fully authenticated `Meteor.user`
        - This is your chance to completely override the user authentication process. If a `user` is
          returned, any `userId` and `token` will be ignored, as it's assumed that you have already
          successfully authenticated the user (by whatever means you deem necessary). The given user
          is simply attached to the [endpoint context](#endpoint-context), no questions asked.
          
      For either level of auth described above, you can optionally return a custom error response by 
      providing that response in an `error` field of your response object. The `error` value can be 
      [any valid response](#response-data). If an `error` field exists in the object returned from 
      your custom auth function, all other fields will be ignored. Do **not** provide an `error` 
      value if you intend for the authentication to pass successfully.

##### `defaultHeaders`
- _Object_
- Default: `{ 'Content-Type': 'application/json' }`
- The response headers that will be returned from every endpoint by default. These can be overridden
  by [returning `headers` of the same name from any endpoint](#response-data).

##### `defaultOptionsEndpoint`
- [_Endpoint_](#endpoint-configuration)
- Default: undefined
- If an endpoint is provided, it will be used as the OPTIONS endpoint on all routes, except those
  that have one manually defined. This can be used to DRY up your API, since OPTIONS endpoints will
  frequently [respond generically](http://zacstewart.com/2012/04/14/http-options-method.html) across
  all routes.

##### `enableCors`
- _Boolean_
- Default: `true`
- If true, enables cross-origin resource sharing ([CORS]). This allows your API to receive requests
  from _any_ domain (when `false`, the API will only accept requests from the domain where the API
  is being hosted. _Note: Only applies to requests originating from browsers)._

##### `onLoggedIn`
- _Function_
- Default: `undefined`
- A hook that runs once a user has been successfully logged into their account via the `/login`
  endpoint. [Context](#endpoint-context) is the same as within authenticated endpoints. Any
  returned data will be added to the response body as `data.extra`.

##### `onLoggedOut`
- _Function_
- Default: `undefined`
- Same as onLoggedIn, but runs once a user has been successfully logged out of their account via
  the `/logout` endpoint. [Context](#endpoint-context) is the same as within authenticated
  endpoints. Any returned data will be added to the response body as `data.extra`.

##### `onAuth`
- _Function_
- Default: `undefined`
- Function called right after authorization has been checked and before the route `action()` is called. This
  can be used to add extra checks to the authorization. 

##### `onAction`
- _Function_
- Default: `undefined`
- Function called right before the `action()` on the route is called. This allows to capture all calls to the api
  and potentially manipulate the endpoint context that is passed to the route.

##### `onReturn`
- _Function_
- Default: `undefined`
- Function called right before the result is returned to the client. This allows to capture the return value
  and potentially manipulate before sending.

##### `prettyJson`
- _Boolean_
- Default: `false`
- If `true`, render formatted JSON in response.

##### `useDefaultAuth`
- _Boolean_
- Default: `false`
- If `true`, `POST /login` and `GET /logout` endpoints are added to the API. See [Authenticating]
  (#authenticating) for details on using these endpoints.

##### `version`
- _String_
- Default: `null`
- URL path versioning is the only type of API versioning currently available, so if a version is
  provided, it's appended to the base path of all routes that belong to that API
    ```javascript
    // Base URL path: my-api/v1/
    ApiV1 = new Restivus({
      apiPath: 'my-api/',
      version: 'v1'
    });

    // Base URL path: my-api/v2/
    ApiV2 = new Restivus({
      apiPath: 'my-api/',
      version: 'v2'
    });
    ```

Here's a sample configuration with the complete set of options: 

**Warning! For demo purposes only - using this configuration is not recommended!**

###### JavaScript
```javascript
  new Restivus({
    apiPath: 'my-api/',
    auth: {
      token: 'auth.apiKey',
      user: function () {
        return {
          userId: this.request.headers['user-id'],
          token: this.request.headers['login-token']
        };
      }
    },
    defaultHeaders: {
      'Content-Type': 'application/json'
    },
    onLoggedIn: function () {
      console.log(this.user.username + ' (' + this.userId + ') logged in');
    },
    onLoggedOut: function () {
      console.log(this.user.username + ' (' + this.userId + ') logged out');
    },
    onAuth: function(auth) {
        // do some extra checks on the `auth.user` object
        return auth;    
    },
    onAction: function() {
        this.extraData = 'some extra data';    
    },
    onReturn: function(returnValue) {
        return returnValue;    
    },
    prettyJson: true,
    useDefaultAuth: true,
    version: 'v1'
  });
```

## Defining Collection Routes

One of the most common uses for a REST API is exposing a set of operations on your collections.
Well, you're in luck, because this is almost _too easy_ with Restivus! All available REST endpoints
(except `patch` and `options`, for now) can be generated for a Mongo Collection using
`Restivus#addCollection()`. This generates two routes by default:

**`/api/<collection>`**
- Operations on the entire collection
-  `GET` and `POST`

**`/api/<collection>/:id`**
- Operations on a single entity within the collection
- `GET`, `PUT`, `PATCH` and `DELETE`

### Collection

The first - and only required - parameter of `Restivus#addCollection()` is a Mongo Collection.
Please check out the [Meteor docs](http://docs.meteor.com/#/full/collections) for more on creating
collections. The `Meteor.users` collection will have [special endpoints]
(#users-collection-endpoints) generated.

### Collection Options

Route and endpoint configuration options are available in `Restivus#addCollection()` (as the 2nd,
optional parameter).

#### Route Configuration

The top level properties of the options apply to both routes that will be generated
(`/api/<collection>` and `/api/<collection>/:id`):

##### `path`
- _String_
- Default: Name of the collection (the name passed to `new Mongo.Collection()`, or `'users'` for
  `Meteor.users`)
- The base path for the generated routes. Given a path `'other-path'`, routes will be generated at
  `'api/other-path'` and `'api/other-path/:id'`

##### `routeOptions`
- _Object_
- `authRequired` _Boolean_
  - Default: `false`
  - If true, all endpoints on these routes will return a `401` if the user is not properly
    [authenticated](#authenticating).
- `roleRequired` _String or Array of Strings or Object_
  - Default: `undefined` (no role required)
  - The acceptable user roles for all endpoints on this route (e.g., `'admin'`, `['admin', 'dev']`).
    Additional role permissions can be defined on specific endpoints. If the authenticated user does
    not belong to at least one of the accepted roles, a `403` is returned. Since a role cannot be
    verified without an authenticated user, setting the `roleRequired` implies `authRequired: true`,
    so that option can be omitted without any consequence. For more on setting up roles, check out
    the [`alanning:roles`][alanning-roles] package.
  - If you need to set the scope (or group) of the roleRequired, set this value to an object like:
  `{roles: 'admin', scope: 'clients'}` or `{roles: ['admin', 'finance'], scope: 'clients'}` 

##### `excludedEndpoints`
- _String or Array of Strings_
- Default: `undefined`
- The names of the endpoints that should _not_ be generated (see the `endpoints` option below for a
  complete list of endpoint names).

##### `endpoints`
- _Object_
- Default: `undefined` (all available endpoints generated)
- Each property of this object corresponds to a REST endpoint. In addition to the
  `excludedEndpoints` list, you can also prevent an endpoint from being generated by setting its
  value to `false`. All other endpoints will be generated. The complete set of configurable
  properties on these endpoints is described in the [Endpoint Configuration](#endpoint-configuration)
  section below. Here is a list of all available endpoints, including their corresponding HTTP method,
  path, and a short description of their behavior:
  - `getAll` [_Endpoint_](#endpoint-configuration)

CHANGELOG.md
# Change Log

## [v2.1.0] - 2023-11-09

#### Added
Added support for Bearer token authentication, when using [leaonline:oauth2-server](https://github.com/leaonline/oauth2-server) 

## [v2.0.0] - 2022-01-18

#### Added
Added support for async API routes

## [v1.1.0] - 2020-12-17

#### Added
onAuth, onAction, onReturn functions to global options.

## [v1.0.0] - 2020-12-10

#### Changed
Decaffeinated project, added ecmascript support, expanded usage of `roleRequired` to allow setting of scope (or group).

## [v0.8.12] - 2017-01-23

#### Added
- Support for SHA-256 hashed passwords sent from client in the [default login endpoint](https://github.com/kahmali/meteor-restivus#logging-in)
- [PATCH support](https://github.com/kahmali/meteor-restivus#patch) for default collection endpoints 

#### Changed
- Allow custom error responses to be returned from [custom auth functions](https://github.com/kahmali/meteor-restivus#auth) by providing an `error` value in return object

#### Fixed 
- Bcrypt build error (required upgrade of `accounts-password` dependency to v1.3.3)


## [v0.8.11] - 2016-06-16

#### Fixed
- Allow null or undefined to be returned from endpoint function when HTTP response is handled manually ([#135][])


## [v0.8.10] - 2016-03-28

#### Changed
- Add result from `onLoggedOut` hook to `data.extra` field ([#196][]) 
- Remove the default `onLoggedIn` and `onLoggedOut` hooks that returned an empty object (docs incorrectly listed them as `undefined`, but the result was not being used prior to [v0.8.9]) 


## [v0.8.9] - 2016-03-14

#### Changed
- Add result from `onLoggedIn` hook to `data.extra` field ([#194][]) 


## [v0.8.8] - 2016-03-10

#### Changed
- Update to the latest version of `simple:json-routes` ([v2.1.0](https://github.com/stubailo/meteor-rest/tree/master/packages/json-routes#210))
  - Resolves errors when using alongside Iron Router ([#185][]) 


## [v0.8.7] - 2016-02-24

#### Fixed
- Declare dependency on accounts-password ([#169][])


## [v0.8.6] - 2016-02-24

#### Changed
- Update to the latest version of `simple:json-routes` ([v2.0.1](https://github.com/stubailo/meteor-rest/tree/master/packages/json-routes#201))
  - Increases max request size to 50mb ([#78][] and [#179][]) 


## [v0.8.5] - 2016-02-24

#### Fixed
- Remove `Route` class from global namespace ([#170][])


## [v0.8.4] - 2015-08-18

#### Added
- Default OPTIONS endpoint when CORS is enabled ([#99][])

#### Fixed
- Enable CORS with default authentication when both are configured ([#99][])


## [v0.8.3] - 2015-08-12

#### Changed
- Accept explicit `username` or `email` in default login endpoint ([#121][])


## [v0.8.2] - 2015-08-05

#### Changed
- Update `simple:json-routes` to v1.0.4 to remove Connect deprecation warnings ([#118][])


## [v0.8.1] - 2015-07-07

#### Added 
- Default logout endpoint at `POST /api/logout`

#### Deprecated
- Default logout endpoint at `GET /api/logout` (#100)
  - Will be removed in v1.0
  - Use `POST /api/logout` instead


## [v0.8.0] - 2015-07-06

**_WARNING!_ API-breaking changes! Please be aware when upgrading!**

_Note: API-breaking changes are in **bold**_

#### Added
- Support [API versioning][versioning an api] via URL path
- API configuration option for defining a default OPTIONS endpoint (used on 
  any route that doesn't define it's own OPTIONS endpoint): `defaultOptionsEndpoint`
    
#### Changed
- **Export Restivus as a "class" instead of an object**
- Replace Iron Router with [JSON Routes]
  (https://github.com/stubailo/meteor-rest/tree/master/packages/json-routes)! ([#38][])
- Return standard response codes in any autogenerated endpoints ([#48][])
  - `201`: Resource created (in `POST` collection endpoints)
  - `403`: Role permission errors
  - `405`: API method not found (but route exists)
- Remove dependency on `useAuth` API config option for endpoint authentication ([#49][])
  
#### Deprecated
- `useAuth` API config option (replaced with `useDefaultAuth` to better reflect it's purpose, which 
  is to generate the default auth endpoints)

#### Fixed
- Iron Router related issues
  - Issue [#24][]: Interference with other routers
  - Issue [#35][]: App hangs on iOS
  - Issue [#43][]: Deployed app shows `iron:router` welcome screen
- Issue [#91][]: Return a `401` error when auth info is not provided in endpoint requiring default 
  auth (previously returned `500` error) 
  
#### Removed
- **`deleteAll` collection endpoint (it had the potential to be quite destructive - [#47][])**
- **`Restivus.configure()` (configuration now available via Restivus constructor)**


## [v0.7.1] - 2015-07-01

#### Fixed
- Issue #95: On 401 and 403 errors, Restivus should wait before sending any data


## [v0.7.0] - 2015-06-18

**_WARNING!_ Potentially breaking changes! Please be aware when upgrading!**

#### Changed
- Update default auth endpoints to match current Accounts token storage ([#79][])
  - **_WARNING!_ All clients consuming a Restivus API _with the default authentication_ will need to 
    reauthenticate after this update**
  - Login token is now stored as `hashedToken` instead of `token`
- Return `401 Unauthorized` for failed authentication
- When logging in with bad credentials, randomly delay the response ([#81][])
- Declare dependency on `accounts-base` package
  

## [v0.6.6] - 2015-05-25
  
#### Changed
- Improve the Contributing Guidelines
  - Tighten up some conventions
  - Add more detailed information

#### Fixed
- Issue [#68][]: Make Restivus a server-only package (`Restivus` is now `undefined` on the client)


## [v0.6.5] - 2015-04-27

#### Added
- Include an endpoint's configuration options [in its context]
  (https://github.com/kahmali/meteor-restivus#thisendpointoption)
- Allow [default response headers](https://github.com/kahmali/meteor-restivus#defaultheaders) to be 
  configured
- Allow CORS to be disabled in API config (using [`enableCors`]
  (https://github.com/kahmali/meteor-restivus#enablecors) option)
    - Resolves Issue [#39][]
    
#### Changed

lib/route.js
import { Meteor } from 'meteor/meteor';
import { DDPCommon } from 'meteor/ddp-common';
import { DDP } from 'meteor/ddp-client';
import { JsonRoutes } from 'meteor/simple:json-routes';
import { loginWithBearerToken } from "./oauth";

export const Route = function (api, path, options, endpoints1) {
    this.api = api;
    this.path = path;
    this.options = options;
    this.endpoints = endpoints1;
    if (!this.endpoints) {
        this.endpoints = this.options;
        this.options = {};
    }
};

const availableMethods = ['get', 'post', 'put', 'patch', 'delete', 'options'];
Route.prototype.addToApi = function () {
    const self = this;

    if (_.contains(this.api._config.paths, this.path)) {
        throw new Error('Cannot add a route at an existing path: ' + this.path);
    }

    this.endpoints = _.extend({
        options: this.api._config.defaultOptionsEndpoint
    }, this.endpoints);
    this._resolveEndpoints();
    this._configureEndpoints();

    this.api._config.paths.push(this.path);

    const allowedMethods = _.filter(availableMethods, function (method) {
        return _.contains(_.keys(self.endpoints), method);
    });

    const rejectedMethods = _.reject(availableMethods, function (method) {
        return _.contains(_.keys(self.endpoints), method);
    });

    const fullPath = this.api._config.apiPath + this.path;
    _.each(allowedMethods, function (method) {
        const endpoint = self.endpoints[method];
        return JsonRoutes.add(method, fullPath, async function (req, res) {
            let responseInitiated = false;
            let doneFunc = function () {
                return responseInitiated = true;
            };

            const endpointContext = Object.assign({
                urlParams: req.params,
                queryParams: req.query,
                bodyParams: req.body,
                request: req,
                response: res,
                done: doneFunc
            }, endpoint);

            let responseData = null;
            try {
                responseData = await self._callEndpoint(endpointContext, endpoint);
            } catch (e) {
                const  env = process.env.NODE_ENV || 'development';
                if (res.statusCode < 400) {
                    res.statusCode = 500;
                }

                if (e.status) {
                    res.statusCode = e.status;
                }

                let msg;
                if (env === 'development') {
                    msg = (e.stack || e.toString()) + '\n';
                } else {
                    msg = 'Server error.';
                }
                console.error(e.stack || e.toString());

                if (res.headersSent) {
                    return req.socket.destroy();
                }

                res.setHeader('Content-Type', 'text/html');
                res.setHeader('Content-Length', Buffer.byteLength(msg));
                if (req.method === 'HEAD') {
                    return res.end();
                }

                res.end(msg);
                return;
            }

            if (responseInitiated) {
                res.end();
                return;
            } else {
                if (res.headersSent) {
                    throw new Error('Must call this.done() after handling endpoint response manually: ' + method + ' ' + fullPath);
                } else if (responseData === null || responseData === void 0) {
                    throw new Error('Cannot return null or undefined from an endpoint: ' + method + ' ' + fullPath);
                }
            }

            if (responseData.body && (responseData.statusCode || responseData.headers)) {
                return self._respond(res, responseData.body, responseData.statusCode, responseData.headers);
            } else {
                return self._respond(res, responseData);
            }
        });
    });

    _.each(rejectedMethods, function (method) {
        JsonRoutes.add(method, fullPath, function (req, res) {
            const responseData = {
                status: 'error',
                message: 'API endpoint does not exist'
            };
            const headers = {
                'Allow': allowedMethods.join(', ').toUpperCase()
            };

            return self._respond(res, responseData, 405, headers);
        });
    });
};

/*
  Convert all endpoints on the given route into our expected endpoint object if it is a bare
  function

  @param {Route} route The route the endpoints belong to
 */
Route.prototype._resolveEndpoints = function () {
    _.each(this.endpoints, function (endpoint, method, endpoints) {
        if (_.isFunction(endpoint)) {
            return endpoints[method] = {
                action: endpoint
            };
        }
    });
};

/*
  Configure the authentication and role requirement on all endpoints (except OPTIONS, which must
  be configured directly on the endpoint)

  Authentication can be required on an entire route or individual endpoints. If required on an
  entire route, that serves as the default. If required in any individual endpoints, that will
  override the default.

  After the endpoint is configured, all authentication and role requirements of an endpoint can be
  accessed at <code>endpoint.authRequired</code> and <code>endpoint.roleRequired</code>,
  respectively.

  @param {Route} route The route the endpoints belong to
  @param {Endpoint} endpoint The endpoint to configure
 */
Route.prototype._configureEndpoints = function () {
    _.each(this.endpoints, (endpoint, method) => {
        if (method !== 'options') {

            if (!_.has(endpoint, 'roleRequired') && _.has(this.options, 'roleRequired')) {
                // set endpoint roleRequired to route roleRequired
                endpoint.roleRequired = this.options.roleRequired;
            }

            if (!_.has(endpoint, 'authRequired')) {
                // check whether global authRequired is on OR roleRequired on the endpoint
                endpoint.authRequired = (
                    (_.has(this.options, 'authRequired') && this.options.authRequired)
                    ||
                    endpoint.roleRequired
                );
            }
        }
    });
};

/*
  Authenticate an endpoint if required, and return the result of calling it

  @returns The endpoint response or a 401 if authentication fails
 */
Route.prototype._callEndpoint = async function (endpointContext, endpoint) {
    let auth = await this._authAccepted(endpointContext, endpoint);
    const onAuth = this.api._config.onAuth;
    if (_.isFunction(onAuth)) {
        auth = onAuth.call(this, auth) || auth;
    }
    if (auth.success) {
        if (this._roleAccepted(endpointContext, endpoint)) {
            const invocation = new DDPCommon.MethodInvocation({
                isSimulation: false,
                setUserId: function () {
                    throw Error('setUserId not implemented');
                },
                unblock: function () {},
                setHttpStatusCode: function (code) {},
                connection: endpointContext.request,
            });

            const endpointInvocation = Object.assign(invocation, endpointContext);

            const onAction = this.api._config.onAction;
            const onReturn = this.api._config.onReturn;
            return DDP._CurrentInvocation.withValue(invocation, Meteor.wrapAsync(async function (done) {
                endpointInvocation.params = Object.assign({},
                    endpointInvocation.urlParams || {},
                    endpointInvocation.queryParams || {},
                    endpointInvocation.bodyParams || {}
                );

                if (_.isFunction(onAction)) {
                    // global onAction function, this allows to capture all api calls, but also to amend the
                    // context before the action function is called on the route
                    await onAction.call(endpointInvocation);
                }

                let returnValue = await endpoint.action.call(endpointInvocation);

                if (_.isFunction(onReturn)) {
                    // global onReturn function, this allows to capture the return value and manupulate before
                    // sending to the client
                    returnValue = await onReturn.call(endpointInvocation, returnValue) || returnValue;
                }

                return done(null, returnValue);
            }));
        } else {
            return {
                statusCode: 403,
                body: {
                    status: 'error',
                    message: 'You do not have permission to do this.'
                }
            };
        }
    } else {
        if (auth.data) {
            return auth.data;
        } else {
            return {
                statusCode: 401,
                body: {
                    status: 'error',
                    message: 'You must be logged in to do this.'
                }
            };
        }
    }
};

/*
  Authenticate the given endpoint if required

  Once it's globally configured in the API, authentication can be required on an entire route or
  individual endpoints. If required on an entire endpoint, that serves as the default. If required
  in any individual endpoints, that will override the default.

  @returns An object of the following format:

      {
        success: Boolean
        data: String or Object
      }

    where `success` is `true` if all required authentication checks pass and the optional `data`
    will contain the auth data when successful and an optional error response when auth fails.
 */
Route.prototype._authAccepted = async function (endpointContext, endpoint) {
    if (endpoint.authRequired) {
        return await this._authenticate(endpointContext);
    } else {
        return {
            success: true
        };
    }
};

/*
  Verify the request is being made by an actively logged in user

  If verified, attach the authenticated user to the context.

  @returns An object of the following format:

      {
        success: Boolean
        data: String or Object
      }

    where `success` is `true` if all required authentication checks pass and the optional `data`
    will contain the auth data when successful and an optional error response when auth fails.
 */
Route.prototype._authenticate = async function (endpointContext) {
    if (endpointContext.request?.headers?.authorization) {
        let auth;
        // oauth authorization header using oauth2 server package
        const header = endpointContext.request.headers.authorization;
        const parts = header.split(' ');
        const scheme = parts[0];
        const credentials = parts[1];
        if (/^Bearer$/i.test(scheme)) {
            try {
                auth = await loginWithBearerToken(credentials);
                return {
                    success: true,
                    data: auth
                };
            } catch (e) {
                console.error(e);
                return {
                    success: false
                };
            }
        }

        return {
            success: false
        };
    }

    const auth = this.api._config.auth.user.call(endpointContext);
    if (!auth) {
        return {
            success: false
        };
    }

    if (auth.userId && auth.token && !auth.user) {
        const userSelector = {};
        userSelector._id = auth.userId;
        userSelector[this.api._config.auth.token] = auth.token;
        auth.user = Meteor.users.findOne(userSelector);
    }

    if (auth.error) {
        return {
            success: false,
            data: auth.error
        };
    }

    if (auth.user) {
        endpointContext.user = auth.user;
        endpointContext.userId = auth.user._id;
        return {
            success: true,
            data: auth
        };
    }

    return {
        success: false
    };
};

/*
  Authenticate the user role if required

  Must be called after _authAccepted().

  @returns True if the authenticated user belongs to <i>any</i> of the acceptable roles on the
           endpoint
 */
Route.prototype._roleAccepted = function (endpointContext, endpoint) {
    if (endpoint.roleRequired) {
        // weak requirement on the roles package, if not installed this will fail
        import { Roles } from 'meteor/alanning:roles';

        let rolesRequired = endpoint.roleRequired;
        if (_.isString(rolesRequired) || _.isArray(rolesRequired)) {
            return Roles.userIsInRole(endpointContext.user._id, rolesRequired);
        } else if (_.isObject(rolesRequired)) {
            check(rolesRequired, {
                roles: Match.OneOf(String, Array),
                scope: Match.Optional(String)
            });
            return Roles.userIsInRole(endpointContext.user._id, rolesRequired.roles, rolesRequired.scope);
        }

        return false;
    }

    return true;
};

/*
  Respond to an HTTP request
 */
Route.prototype._respond = function (response, body, statusCode, headers) {
    if (statusCode == null) {
        statusCode = 200;
    }

    if (headers == null) {
        headers = {};
    }

    const defaultHeaders = this._lowerCaseKeys(this.api._config.defaultHeaders);
    headers = this._lowerCaseKeys(headers);
    headers = _.extend(defaultHeaders, headers);
    if (headers['content-type'].match(/json|javascript/) !== null) {
        if (this.api._config.prettyJson) {
            body = JSON.stringify(body, void 0, 2);
        } else {
            body = JSON.stringify(body);
        }
    }

    const sendResponse = function () {
        response.writeHead(statusCode, headers);
        response.write(body);
        return response.end();
    };

    if (statusCode === 401 || statusCode === 403) {
        const minimumDelayInMilliseconds = 500;
        const randomMultiplierBetweenOneAndTwo = 1 + Math.random();
        const delayInMilliseconds = minimumDelayInMilliseconds * randomMultiplierBetweenOneAndTwo;
        return Meteor.setTimeout(sendResponse, delayInMilliseconds);
    }

    return sendResponse();
};

/*
  Return the object with all of the keys converted to lowercase
 */
Route.prototype._lowerCaseKeys = function (object) {
    return _.chain(object)
        .pairs()
        .map(function (attr) {
            return [attr[0].toLowerCase(), attr[1]];
        })
        .object()
        .value();
};
